naga/src/back/glsl/features.rs

```diff
@@ -469,6 +477,48 @@ impl<'a, W> Writer<'a, W> {
                         }
                     }
                 }
+                Expression::ImageSample { image, level, offset, .. } => {
+                    if let TypeInner::Image {
+                        dim,
+                        arrayed,
+                        class: ImageClass::Depth { .. },
+                    } = *info[image].ty.inner_with(&module.types) {
+                        let lod = matches!(level, SampleLevel::Zero | SampleLevel::Exact(_));
+                        let bias =  matches!(level, SampleLevel::Bias(_));
+                        let cube =  dim == ImageDimension::Cube;
+                        let array2d = dim == ImageDimension::D2 && arrayed;
+
+                        // We have an workaround for cases other than samplerCubeArrayShadow,
+                        // so we don't *need* this extension for those cases.
+                        // But if we we're explicitly allowed to use it (`WriterFlags::TEXTURE_SHADOW_LOD`),
+                        // we prefer it over the workaround.
+                        let use_grad_workaround = !(
+                            self.options.writer_flags.contains(WriterFlags::TEXTURE_SHADOW_LOD)
+                            || (cube && arrayed)
+                        );
```

Comment:
I find this a bit hard to wrap my head around when combined with the other operations below.

It would be easier to follow if we do something like:
```rust
let possibly_needs_grad_workaround = (sampler2DArrayShadow || samplerCubeShadow) && crate::SampleLevel::Zero;
is_used |= possibly_needs_grad_workaround && self.options.writer_flags.contains(WriterFlags::TEXTURE_SHADOW_LOD);
```

Reply:
I hated this code too, clippy forced me to invert the conditon...
Should be more readable now.